### ISSUE TITLE
[Research] Evaluate recovery-state training for bounded LIBERO Repair

### ISSUE THREAD

**Issue by pome223** [pome223/missionos#85]
## Status

This is a separate research proposal, not an implementation commitment and not
a claim that additional training will produce stable Repair.

Issue #75 established bounded checkpoint-specific executor differences and a
backend-neutral five-axis diagnostic contract. It also established that a
privileged registered push skill can reach the missing predicate yet still fail
preservation. This issue isolates the remaining training-data question instead
of mixing it with checkpoint integration and runtime governance.

## Research question

Does training on explicit disturbed-state-to-corrective-action transitions
improve **stable, preservation-constrained** LIBERO Repair relative to nominal
training alone?

## Required data contract

Each recovery example must bind:

- source simulator-state digest and held-out split identity;
- actual source predicate vector and preserved-object reference poses;
- observation and proprioception schema digest;
- corrective action sequence and applied-action digest;
- actual per-step predicate and preservation traces;
- stable-hold result;
- whether privileged state was used to generate or label the trajectory.

Failure rollouts without corrective transitions are not recovery
demonstrations and must be labeled separately.

## Evaluation

Use the existing five-axis contract:

1. action activity;
2. corrective alignment;
3. predicate recovery;
4. preservation;
5. stable hold.

Compare nominal-only and nominal-plus-recovery training under matched model,
optimizer, compute, seeds, observations, and action representation. Keep task,
fixture, and seed cohorts separate. A first predicate conjunction is not stable
Repair.

## Gates and stop rules

- No paid training before a written dataset-size, GPU-hour, cost-cap, and
  teardown estimate is reviewed.
- No evaluation fixture or action trace may enter training data.
- No simulator-state-privileged label is presented as observation-only skill.
- No recovery-rate claim from one task, fixture, seed, or checkpoint.
- Stop if the nominal control gate, adapter parity, or exact-fixture oracle gate
  fails.
- Generated futures and model self-reports are never verifier evidence.

## Completion criteria

- [ ] Publish the recovery-transition data schema and leakage audit.
- [ ] Record a bounded compute/cost go-no-go decision.
- [ ] Pre-register model cells, seeds, fixtures, action budgets, and stable-hold
      criteria.
- [ ] If approved, train the minimum matched comparison.
- [ ] Publish positive or negative five-axis results and delete transient paid
      resources.

## Explicit non-claims

This issue does not establish recovery-trained executor capability, physical
execution, controller ACK, real-world safety, or a general VLA/WAM comparison.

**Comment by pome223** [comment]
Phase 0 implementation is in PR #87: https://github.com/pome223/missionos/pull/87

The branch now contains:

- `missionos.libero_recovery_transition.v1` required-field and provenance validation;
- fail-closed exclusion of evaluation states, fixtures, action traces, and evaluation-split records from training;
- a matched `nominal_only` versus `nominal_plus_recovery` preregistration;
- the existing five-axis contract with a required 20-step stable hold;
- a digest-bound Phase 0 record whose current decision is `NO_GO`.

Observed CLI result:

```text
status=valid
decision=NO_GO
paid_training_authorized=false
gpu_provision_authorized=false
training_performed=false
```

Validation: 6 focused tests passed; the full contract suite passed with 2273 tests and 6 warnings; Ruff, JSON parsing, and diff checks passed.

The `NO_GO` is the intended fail-closed result. There is no admitted training manifest, selected trainable checkpoint, measured dataset size, GPU-hour estimate, reviewed cost cap, or reviewed automatic teardown plan yet. No cloud resource was created and no paid training was authorized. PR merge would complete only the schema/audit design, readiness decision, and preregistration items; the matched training run remains separately approval-gated.

**Comment by pome223** [comment]
Phase 0 is now merged in #87 and #88. The concrete preregistration gap found after #87 was corrected before completion: task/fixture partitions, environment and model seeds, disturbances, the 128-action policy ceiling, and the 20-step hold are now fixed and fail closed on overlap or budget drift. Both CI jobs passed on both PRs. The issue remains open at NO-GO; paid training and GPU provisioning are still unauthorized.

**Comment by pome223** [comment]
Phase 1 raw transition capture is now in PR #90.

Bounded result on a preregistered non-evaluation condition (`init=0`, `seed=101`, 3 cm):

- scripted oracle recovery established in 52 applied actions
- complete predicate vector held for 20 steps
- no preservation violation observed
- all 52 pre-action samples captured as agentview RGB, wrist RGB, 8-D state, actual 7-D action, and 20 Hz timestamp
- real LIBERO simulator boundary exercised

Cost controls were CPU-only, one-hour max duration with automatic DELETE and restart disabled. The capture VM ran for under 11 minutes; the VM and boot disk were both confirmed absent afterward. No GPU was provisioned.

This advances data-generation feasibility only. The artifact is still `raw_transition_capture_complete_conversion_and_admission_pending`; LeRobot v2 MP4/Parquet conversion, multi-example manifest, leakage audit, GPU-hour measurement, and a reviewed hard cost cap remain open. Paid training and GPU provisioning remain NO-GO.

**Comment by pome223** [comment]
PR #91 is merged and closes the first structural conversion gate.

The 52-step raw candidate from PR #90 now converts deterministically to:

- two AV1 video streams, 256 x 256, 20 Hz, 52 frames each;
- one 52-row Parquet episode with fixed-size float32 state `[8]` and action `[7]`;
- LIBERO-compatible `info.json`, `modality.json`, `episodes.jsonl`, and `tasks.jsonl`.

Two independent local conversions produced identical hashes for every output file. This used local CPU only and created no cloud resources.

The status remains `lerobot_v2_conversion_complete_admission_pending`. The remaining pre-training work is to generate a small multi-condition cohort, compute dataset statistics, materialize exact holdout exclusions, pass the leakage audit, and measure the resulting dataset/GPU-hour estimate. Paid training and GPU provisioning remain NO-GO.

**Comment by pome223** [comment]
PR #92 completed the bounded dataset-admission gate.

- Training candidates: 4/4 privileged-oracle recoveries, 206 aligned frames, all converted to LeRobot v2.
- Evaluation-only holdouts: preregistered `(init, seed)=(4,0),(12,0),(15,0)`, each exact state/fixture/action identity recorded with `used_for_training=false`.
- Leakage audit: passed across four training records and six evaluation references with zero collisions.
- Admission: four recovery candidates admitted for a future matched experiment.
- Runtime verification: 2466 tests passed locally; Python 3.11 and 3.13 CI passed.
- Cost boundary: CPU-only temporary VMs were deleted with their boot disks; no GPU, paid training, or model inference was invoked.

Issue #85 remains open. Paid training is still **NO-GO** until a trainable checkpoint is pinned, minimum GPU-hours are estimated, a reviewed hard currency cost cap and automatic teardown plan exist, and a separate human approval is recorded. Dataset admission is not training authorization.

**Comment by pome223** [comment]
PR #93 merged the read-only GR00T training cost preflight.

Current result remains **NO-GO** for paid training:

- pinned research candidate: `nvidia/GR00T-N1.7-3B@2fc962b973bccdd5d8ce4f67cc63b264d6886495`
- checkpoint license recorded as research/evaluation-only; commercial use and derivative-weight publication remain unauthorized
- dependent `nvidia/Cosmos-Reason2-2B` is auto-gated and current access is unverified
- NVIDIA minimum fine-tuning VRAM is 40 GB; the available L4 is 24 GB and rejected
- observed `us-central1` A100 quota is zero
- proposed cost boundary is 5,000 JPY total, 3,500 JPY prelaunch estimate ceiling, one GPU, 60-minute maximum runtime, termination `DELETE`, boot-disk auto-delete, and no persistent model cache
- the proposal is not yet approved, and billing alerts are explicitly not treated as a hard stop

Validation: 2472 tests passed locally; Python 3.11 and 3.13 CI passed. No VM, GPU, quota request, model download, training, or inference was started.

The next possible billable step, only after separate review and approval, is a resumable 200-step throughput calibration rather than the matched experiment.

**Comment by pome223** [comment]
PR #94 merged the next cost-bounded step.

- Submitted a quota preference for one NVIDIA A100 in us-central1. Current state is preferred=1, granted=0, reconciling=true; this created no compute cost.
- Read-only Hugging Face access probing found no configured local token. GR00T checkpoint weight access remains unverified, and the Cosmos backbone remains gated/unverified. No model files were downloaded.
- Ran a CPU-only LIBERO geometry screen on one init state and seed. Seven artifact-backed 3 cm / 5 cm conditions produced zero new recoverable training candidates.
- One 5 cm negative-y fixture produced a 128-frame raw negative transition capture, but the scripted oracle did not establish stable recovery in 128 actions. It is not admitted training data.
- The temporary e2-standard-4 VM had a 2700-second maximum duration, no automatic restart, termination DELETE, and boot-disk auto-delete. The instance and disk were both verified absent after the run. No GPU, training, or model inference was invoked; billing total is not yet confirmed.
- Validation: 2475 tests passed locally; Python 3.11 and 3.13 CI passed.

The paid-training decision remains **NO-GO**. The next safe work is only read-only monitoring and access preparation. Do not provision a GPU until all of these are true: authenticated access to both pinned model repositories is verified, the research-only license is reviewed for the intended use, one A100 is actually granted, and the 5,000 JPY proposal cap / 3,500 JPY prelaunch ceiling is reviewed against a fresh price quote. If those gates pass, request separate approval for one resumable 200-step calibration with one GPU and automatic deletion; do not launch the matched experiment directly.

**Comment by pome223** [comment]
PR #95 merged the no-cost access/quota refresh.

Resolved gate:
- the credential already held in Google Cloud Secret Manager can access metadata and issue positive HEAD requests against weight objects at both pinned revisions: GR00T N1.7 and Cosmos Reason2 2B
- no credential value was emitted and no model payload was downloaded

Remaining blockers:
- A100 quota is still preferred=1, granted=0, reconciling=true
- the pinned GR00T repository has conflicting license signals: its LICENSE file limits the work and derivatives to non-commercial research/evaluation, while the same revision README links the current NVIDIA Open Model License and says commercial/non-commercial use is supported; no automated legal conclusion was made
- the 5,000 JPY proposal cap and current price still require review before any billable launch

Cost inventory:
- instances=0, persistent disks=0, snapshots=0
- two prior custom runtime images remain, about 104.82 GiB compressed in total
- at the observed USD 0.000068493/GiB-hour list price, this is approximately USD 5.24 per 730-hour month; this is an estimate, not the billing total
- the images were not deleted because retention versus rebuild is a separate decision

Validation: 2476 tests passed locally; Python 3.11 and 3.13 CI passed.

Decision remains **NO-GO** for GPU provisioning and paid training. The next action requiring user judgment is the pinned GR00T license conflict. The next cloud action is read-only quota monitoring; a grant alone will not authorize a GPU launch.
